man/nvmec.1
.\" Text automatically generated by txt2man
.TH untitled  "19 June 2023" "" ""
.SH NAME
\fBnvmec \fP- No short description
.SH SYNOPSIS
.nf
.fam C
\fBnvmec\fP <command> [<args>]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
No long description
.SH COMMANDS
.TP
.B
\fBnvmec-show\fP(1)
Print the given NVMe command-file
.TP
.B
\fBnvmec-create\fP(1)
Create an NVMe command-file
.RE
.PP

.SH OPTIONS
\fB--help\fP
Print the synopsis and exit
.SH EXAMPLES
Read the man page for each <command> or consult the command-line \fB--help\fP:
.PP
.nf
.fam C
    $ nvmec <command> --help

.fam T
.fi
.SH SEE ALSO
Full documentation at: <https://xnvme.io/>
.SH AUTHOR
Written by Simon A. F. Lund <simon.lund@samsung.com> on behalf of Samsung


CHANGELOG.rst
.. SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
..
.. SPDX-License-Identifier: BSD-3-Clause

Changelog
=========

The repository is tagged using semantic versioning, e.g. ``v1.2.3``.

The ``next`` branch contains a preview of the upcoming release. The history on
``next`` is subject to change. Upon release, ``next`` is merged with ``main``
and  and tagged ``vX.Y.Z``.

Changes are described in this file in a section matching the version tag.

Known Issues
------------

See the file named ``ISSUES`` in the root of the repository.

v0.7.0
------

This release got a whole lot bigger than it ought to. Aiming for more
frequent releases in the future. A bunch of changes to API,
implementation, and even the license.

Regarding the license. For compatibility with GPL-licensed software.
That is, consumption of xNVMe from GPL-licences software, then xNVMe is
changed from APACHE to BSD-3-Clause. To better determine the license,
the project has aligned with the golden practice of REUSE-compliance
(https://reuse.software/).

* API
  * Refactored to provide a single "entry-point" header
  * Headers no longer include other headers (except for the entry-point)
  * Renamed the command-line api from xnvmec to xnvme_cli

* API / Command-Sets
  * Support for FDP and DSM
  * Support for Key-Value SSDs

* Backends
  * The POSIX backend is replaced by CBI -- Common Backend Implementation
  * Instead of "POSIX" then a handful of commonly useful implementations
    are provided, being "strictly" POSIX was an impractical restriction,
    such as lacking support for features which are provided on modern
    operating systems but not defined by POSIX.

* Backends: FreeBSD
  * Support for kqueue based aio

* Backends: SPDK
  * Now has options to control Admin and I/O command timeout
  * Removed patches no longer needed (available upstream)

* Backends: vfio
  * Removed as subproject and instead links "properly" with the
    system-provided libvfn

* tools
  * switched fio from "external" to "internal" / "upstream" fio engine

For additional details then please have a look at the commit-history,
the v0.7.0 related issue on GitHUB and PR.

v0.6.0
------

A handful of improvements for Windows, additional steps toward removing
third-party vendoring / bundling, addition of experimental tunable knobs for
io_uring/SQPOLL, and several improvements to testing and CI infra.

* API
  - Removed SLIST from API, although "sys/queue.h" are commonly available on
    Linux/FreeBSD, then they are not part of toolchain on Windows.

* Third-party
  - Bumped SPDK to v22.09, and with that removed mutliple out-of-tree patches
    for DPDK.

* CLI
  - The xNVMe command-line library (libxnvmec) and all the cli-tools using it
    are refactored to use common sets of command-line arguments. Along with
    this came a consistent set of CLI-arguments for admin/sync/async.

* Backends
  - ramdisk: The ramdisk got support for write-zeroes, iovec payloads and added
    to CI testing.
  - linux: support for a buffer-allocator using HUGEPAGES and tunable knobs for
    controlling the behavior of io_uring SQPOLL via environment variables.
  - windows: support for the experimental IORING Windows SPDK API and support
    for block devices (SCSI and SATA).
  - spdk: when controllers are re-used for device-handles, events are
    processed as a means to check whether the controller is still "alive"

* CI
  - scan-build now runs for each PR
  - basic tests are now running post-building testing using the RAMDISK
  - Testing of fabrics with TCP transport is now part of the setup

v0.5.0
------

A bit of expansion in the application of xNVMe with support for macOS, a
ramdisk backend, revival of Python language bindings and a refresh of the docs
on NVMe-over-Fabrics.

* API
  - Removed helpers for SGL
  - Add 'subnqn' to 'xnvme_ident', useful for NVMe-oF
  - Add 'hostnqn' to 'opts', useful for NVMe-oF
  - Add support directive-receive and write-with-directives
  - Spec adjusted for NVMe 2.0, still more work needed in this area

* Third-party
  - Bumped fio to v3.32
  - Bumped SPDK to v22.05
  - Removed liburing, now relies on on-system library instead of
    vendoring/bundling, documentation is updated to assist with library
    installation

* Tooling
  - Re-working testing using cijoe 0.9+, that is, switching to CIJOE/pytest for
    testing an CIJOE workflows for instrumentation
  - liburing is no longer bundled with xNVMe, that is, xNVMe now links with
    liburing as discovered on the system via pkg-config. This is done to avoid
    symbol collisions for applications linking or loading liburing and xNVMe.

* be:linux:async:libaio
  - When 'opts.poll_io' is set then poke() will return immmediatly and now wait
    for completions. This allows the use of trading CPU for more IOPS and lower
    per command latency.

* Additional user-space NVMe driver support via libvfn
  - Added 'be:vfio' providing another user-space driver via libvfn

* Preliminary support for macOS
  - Initial implementation using the "core" I/O mechanisms of sync-io, async
    emulation and the threadpool
  - Does enumeration of NVMe devices through the limited interface provided for
    user-space by the macOS kernel
  - Utilizes what is avaiable for admin-command submision

* Prelimiary support for a "ramdisk" device
  - be:ramdisk: added a backend mimicing the behavior of an NVMe NVM namespace
  - Intended as a test-vehicle providing a "device" without requiring anything
    but the consumption of main memory of the system
  - I/O is "stored" using main-memory

* Revival of the xNVMe Python language bindings
  - A re-introduction of the Python bindings, these are now generated and thus
    provide access to the entire xNVMe C API
  - They are added to the testing infrastructure ensuring that they are aligned
    with the library
  - In addition to simple ctypes bindings, then cython headers and bindings
    based on Cython are provided

* Documentation

tools/nvmec.c
// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
//
// SPDX-License-Identifier: BSD-3-Clause

#include <errno.h>
#include <libxnvme.h>

static int
sub_show(struct xnvme_cli *cli)
{
	struct xnvme_spec_cmd cmd = {0};
	int err;

	err = xnvme_buf_from_file(&cmd, sizeof(cmd), cli->args.cmd_input);
	if (err) {
		xnvme_cli_perr("xnvme_buf_from_file()", err);
		xnvme_cli_pinf("Error reading: '%s'", cli->args.cmd_input);
		return err;
	}

	xnvme_spec_cmd_pr(&cmd, XNVME_PR_DEF);

	return 0;
}

static int
sub_create(struct xnvme_cli *cli)
{
	struct xnvme_spec_cmd cmd = {0};
	uint32_t *cdw = (void *)&cmd;
	int err;

	if (cli->given[XNVME_CLI_OPT_NSID]) {
		cmd.common.nsid = cli->args.nsid;
	}

	if (cli->given[XNVME_CLI_OPT_OPCODE]) {
		cmd.common.opcode = cli->args.opcode;
	}

	for (int oi = XNVME_CLI_OPT_CDW00; oi <= XNVME_CLI_OPT_CDW15; ++oi) {
		if (!cli->given[oi]) {
			continue;
		}

		switch (oi) {
		case XNVME_CLI_OPT_CDW00:
			cdw[0] = cli->args.cdw[0];
			break;
		case XNVME_CLI_OPT_CDW01:
			cdw[1] = cli->args.cdw[1];
			break;
		case XNVME_CLI_OPT_CDW02:
			cdw[2] = cli->args.cdw[2];
			break;
		case XNVME_CLI_OPT_CDW03:
			cdw[3] = cli->args.cdw[3];
			break;
		case XNVME_CLI_OPT_CDW04:
			cdw[4] = cli->args.cdw[4];
			break;
		case XNVME_CLI_OPT_CDW05:
			cdw[5] = cli->args.cdw[5];
			break;
		case XNVME_CLI_OPT_CDW06:
			cdw[6] = cli->args.cdw[6];
			break;
		case XNVME_CLI_OPT_CDW07:
			cdw[7] = cli->args.cdw[7];
			break;
		case XNVME_CLI_OPT_CDW08:
			cdw[8] = cli->args.cdw[8];
			break;
		case XNVME_CLI_OPT_CDW09:
			cdw[9] = cli->args.cdw[9];
			break;
		case XNVME_CLI_OPT_CDW10:
			cdw[10] = cli->args.cdw[10];
			break;
		case XNVME_CLI_OPT_CDW11:
			cdw[11] = cli->args.cdw[11];
			break;
		case XNVME_CLI_OPT_CDW12:
			cdw[12] = cli->args.cdw[12];
			break;
		case XNVME_CLI_OPT_CDW13:
			cdw[13] = cli->args.cdw[13];
			break;
		case XNVME_CLI_OPT_CDW14:
			cdw[14] = cli->args.cdw[14];
			break;
		case XNVME_CLI_OPT_CDW15:
			cdw[15] = cli->args.cdw[15];
			break;
		}
	}

	err = xnvme_buf_to_file(&cmd, sizeof(cmd), cli->args.cmd_output);
	if (err) {
		xnvme_cli_perr("xnvme_buf_to_file()", err);
		xnvme_cli_pinf("Error writing: '%s'", cli->args.cmd_output);
		return err;
	}

	if (cli->args.verbose) {
		xnvme_spec_cmd_pr(&cmd, XNVME_PR_DEF);
	}

	return 0;
}

//
// Command-Line Interface (CLI) definition
//
static struct xnvme_cli_sub g_subs[] = {
	{
		"show",
		"Print the given NVMe command-file",
		"Print the given NVMe command-file",
		sub_show,
		{
			{XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},
			{XNVME_CLI_OPT_CMD_INPUT, XNVME_CLI_LREQ},
		},
	},
	{
		"create",
		"Create an NVMe command-file",
		"Create an NVMe command-file",
		sub_create,
		{
			{XNVME_CLI_OPT_NON_POSA_TITLE, XNVME_CLI_SKIP},
			{XNVME_CLI_OPT_CMD_OUTPUT, XNVME_CLI_LREQ},
			{XNVME_CLI_OPT_CDW00, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW01, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW02, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW03, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW04, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW04, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW05, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW06, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW07, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW08, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW09, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW10, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW11, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW12, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW13, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW14, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_CDW15, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_OPCODE, XNVME_CLI_LOPT},
			{XNVME_CLI_OPT_NSID, XNVME_CLI_LOPT},
		},
	},
};

static struct xnvme_cli g_cli = {
	.title = "nvmec: NVMe Command Builder",
	.descr_short = ""
		       "Construct, and show,  NVMe Command files for use with"
		       "e.g.the xNVMe CLI",
	.descr_long = "",
	.subs = g_subs,
	.nsubs = sizeof g_subs / sizeof(*g_subs),
};

int
main(int argc, char **argv)
{
	return xnvme_cli_run(&g_cli, argc, argv, XNVME_CLI_INIT_NONE);
}
